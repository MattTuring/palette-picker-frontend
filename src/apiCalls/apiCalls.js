export const getProjects = async () => {
  const url = 'https://palette-pick-backend.herokuapp.com/api/v1/projects';
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error('There was an error getting projects.');
  }
  const projects = await response.json();
  return projects
}

export const getPalettes = async (id) => {
  const url = `https://palette-pick-backend.herokuapp.com/api/v1/projects/${id}/palettes`
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error('There was an error getting palettes.');
  }
  const palettes = await response.json();
  return palettes
}

export const postProject = async (project) => {
  const url = 'https://palette-pick-backend.herokuapp.com/api/v1/projects';
  const options = {
    method: 'POST',
    body: JSON.stringify(project),
    headers: {
      'Content-Type': 'application/json'
    }
  }
  const response = await fetch(url, options);

  if (!response.ok) {
    throw new Error('There was an error posting project.');
  }

  const result = await response.json();
  return result;
}

export const postPalette = async (palette) => {
  const url = 'https://palette-pick-backend.herokuapp.com/api/v1/palettes';
  const options = {
    method: 'POST',
    body: JSON.stringify(palette),
    headers: {
      'Content-Type': 'application/json'
    }
  }
  const response = await fetch(url, options);

  if (!response.ok) {
    throw new Error('There was an error posting palette.');
  }

  const result = await response.json();
  return result;
}

export const updateProject = async (project, id) => {
  const url = `https://palette-pick-backend.herokuapp.com/api/v1/projects/${id}`;
  const options = {
    method: 'PUT',
    body: JSON.stringify(project),
    headers: {
      'Content-Type': 'application/json'
    }
  }
  const response = await fetch(url, options);

  if (!response.ok) {
    throw new Error('There was an error updating project.');
  }

  const result = await response.json();
  return result;
}

export const updatePalette = async (palette, id) => {
  const url = `https://palette-pick-backend.herokuapp.com/api/v1/palettes/${id}`;
  const options = {
    method: 'PUT',
    body: JSON.stringify(palette),
    headers: {
      'Content-Type': 'application/json'
    }
  }
  const response = await fetch(url, options);

  if (!response.ok) {
    throw new Error('There was an error updating palette.');
  }

  const result = await response.json();
  return result;
}

export const deleteProject = async (id) => {
  const url = `https://palette-pick-backend.herokuapp.com/api/v1/projects/${id}`;
  const options = {
    method: 'DELETE',
    headers: {
      'Content-Type': 'application/json'
    }
  }
  const response = await fetch(url, options);

  if (!response.ok) {
    throw new Error('There was an error deleting project.');
  }

  const result = await response.json();
  return result;
}

export const deletePalette = async (id) => {
  const url = `https://palette-pick-backend.herokuapp.com/api/v1/palettes/${id}`;
  const options = {
    method: 'DELETE',
    headers: {
      'Content-Type': 'application/json'
    }
  }
  const response = await fetch(url, options);

  if (!response.ok) {
    throw new Error('There was an error deleting palette.');
  }

  const result = await response.json();
  return result;
}
