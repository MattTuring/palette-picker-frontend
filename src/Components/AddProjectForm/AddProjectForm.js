import React, { useState } from 'react';
import { postProject } from '../../apiCalls/apiCalls';
import add from '../../add.svg';

const AddProjectForm = ({projects, setProjects}) => {
  const [projectName, setProjectName] = useState('');
  const [opened, setOpened] = useState(false);
  const [notification, setNotification] = useState('');
  const changeInput = event => setProjectName(event.target.value);

  const submitProject = async (event) => {
    event.preventDefault();

    if (projectName === '') {
      return setNotification('There is no name entered!');
    }
    
    if (projects.find(project => project.name === projectName)) {
      return setNotification('Project name must be unique!');
    }

    try {
      const project = await postProject({name: projectName});
      setProjects([...projects, project]);
      setOpened(!opened);
    }
    catch (error) {
      console.error(error);
    }
  }

  const showForm = () => setOpened(!opened);

  return (
    <section className="new-project-form">
      <h3 onClick={showForm}>
        {!opened && <img src={add} alt="add" />}
        Add new project
      </h3>
      {notification !== '' && <p className="error-not">{notification}</p>}
      {opened &&
        <form>
          <input
            className="project-name"
            onChange={changeInput}
            placeholder="Name"
            type="text"
            value={projectName} />
          <button
            className="submit-btn"
            onClick={submitProject}>Add Project</button>
        </form>
      }
    </section>
  )
};

export default AddProjectForm;
