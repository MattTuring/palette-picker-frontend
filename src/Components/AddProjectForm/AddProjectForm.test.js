import React from 'react';
import { shallow } from 'enzyme';
import AddProjectForm from './AddProjectForm';
import { postProject } from '../../apiCalls/apiCalls';

jest.mock('../../apiCalls/apiCalls')

describe('AddProjectForm', () => {
  const setProjects = jest.fn();

  const mockProps = {
    projects: [{
      id: 1,
      name: "Project"
    }],
    setProjects
  };

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <AddProjectForm {...mockProps} />);
  });

  it('should match the snapshot with all data passed in correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the snapshot after clicking on header', () => {
    wrapper.find('h3').simulate('click');
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the snapshot after clicking on header and submitting empty form', () => {
    wrapper.find('h3').simulate('click');
    wrapper.find('.submit-btn').simulate('click', {
      preventDefault: jest.fn()
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('should call postProject after click on ', () => {
    wrapper.find('h3').simulate('click');
    wrapper.find('input').simulate('change', {
      target: {value: 'Project'}
    });
    wrapper.find('.submit-btn').simulate('click', {
      preventDefault: jest.fn()
    });
    expect(postProject).toHaveBeenCalled();
  });

  it('should call setProjects after click on submit button', () => {
    wrapper.find('h3').simulate('click');
    wrapper.find('input').simulate('change', {
      target: {value: 'Project'}
    });
    wrapper.find('.submit-btn').simulate('click', {
      preventDefault: jest.fn()
    });
    expect(setProjects).toHaveBeenCalled();
  });
});
