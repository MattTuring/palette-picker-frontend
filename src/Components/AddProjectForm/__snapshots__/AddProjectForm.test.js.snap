// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddProjectForm should match the snapshot after clicking on header 1`] = `
<section
  className="new-project-form"
>
  <h3
    onClick={[Function]}
  >
    Add new project
  </h3>
  <form>
    <input
      className="project-name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <button
      className="submit-btn"
      onClick={[Function]}
    >
      Add Project
    </button>
  </form>
</section>
`;

exports[`AddProjectForm should match the snapshot after clicking on header and submitting empty form 1`] = `
<section
  className="new-project-form"
>
  <h3
    onClick={[Function]}
  >
    Add new project
  </h3>
  <p
    className="error-not"
  >
    There is no name entered!
  </p>
  <form>
    <input
      className="project-name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <button
      className="submit-btn"
      onClick={[Function]}
    >
      Add Project
    </button>
  </form>
</section>
`;

exports[`AddProjectForm should match the snapshot with all data passed in correctly 1`] = `
<section
  className="new-project-form"
>
  <h3
    onClick={[Function]}
  >
    <img
      alt="add"
      src="add.svg"
    />
    Add new project
  </h3>
</section>
`;
